central/resourcecollection/datastore/datastore_impl.go

```diff
@@ -123,40 +162,177 @@ func (ds *datastoreImpl) addCollectionToGraph(obj *storage.ResourceCollection) e
+	}
+
+	// determine deletions, any remaining edge in curChildren should be removed
+	for key := range curChildren {
+		ret.removedChildrenIDs = append(ret.removedChildrenIDs, key)
+	}
+
+	// since adding can cause cycles but removing never can, we remove first
+	for _, childID := range ret.removedChildrenIDs {
+		err = graphCopy.DeleteEdge(obj.GetId(), childID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// add new edges
+	for _, childID := range ret.addedChildrenIDs {
+		err = graphCopy.AddEdge(obj.GetId(), childID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// if this is a dryrun we just return and don't need to pass the operation object
+	if dryrun {
+		return nil, nil
+	}
+
+	// operation succeeded, update graph
+	ds.graph = graphCopy
+
+	return ret, nil
+}
+
+func (ds *datastoreImpl) undoEdgeUpdatesInGraph(updates *graphEdgeUpdates) error {
+	if updates == nil {
+		return errors.New("passed object must be non nil")
+	}
+
 	ds.graphLock.Lock()
 	defer ds.graphLock.Unlock()
 
+	graphCopy, err := ds.graph.Copy()
+	if err != nil {
+		return err
+	}
+
+	// deletions will never cause cycles so do those first
+	for _, addedChildId := range updates.addedChildrenIDs {
+		err = graphCopy.DeleteEdge(updates.parentID, addedChildId)
+		if err != nil {
+			return errors.Wrap(err, "failed to remove added edge")
+		}
+	}
+
+	// add back removed edges
+	for _, removedChildID := range updates.removedChildrenIDs {
+		err = graphCopy.AddEdge(updates.parentID, removedChildID)
+		if err != nil {
+			return errors.Wrap(err, "failed to restore removed edge")
+		}
+	}
+
+	// update the graph
+	ds.graph = graphCopy
+
+	return nil
+}
+
+func (ds *datastoreImpl) deleteCollectionFromGraph(id string) error {
+	ds.initGraphOnce()
+
+	// this function covers removal of the vertex and any edges to or from other vertices, it is also threadsafe
 	return ds.graph.DeleteVertex(id)
 }
 
 func (ds *datastoreImpl) dryRunAddCollectionInGraph(obj *storage.ResourceCollection) error {
 	ds.initGraphOnce()
+
+	var err error
+	dryrunID := "dryrun"
+
+	if err = verifyCollectionObjectNotEmpty(obj); err != nil {
+		return err
+	}
+
+	if obj.GetId() != "" {
+		return errors.New("invalid argument, added collection must not have a preset `id`")
+	}
+
 	ds.graphLock.Lock()
 	defer ds.graphLock.Unlock()
 
-	var ret error
-
 	// we essentially just add the obj to the graph and delete it before returning
-	err := ds.graph.AddVertexByID(obj.GetId(), obj.GetName())
+	dryrunID, err = ds.graph.AddVertex(graphEntry{
+		id: dryrunID,
+	})
 	if err != nil {
 		return err
 	}
 	for _, peer := range obj.GetEmbeddedCollections() {
-		err = ds.graph.AddEdge(obj.GetId(), peer.GetId())
+		err = ds.graph.AddEdge(dryrunID, peer.GetId())
 		if err != nil {
-			ret = err
 			break
 		}
 	}
 
-	err = ds.graph.DeleteVertex(obj.GetId())
-	if err != nil && ret == nil {
-		ret = err
+	deleteErr := ds.graph.DeleteVertex(dryrunID)
+	if err != nil {
+		return err
 	}
-
-	return ret
+	return deleteErr
```

Comment:
This is pretty much the same as `updateCollectionInGraphWorkflow` except that there are always no edges to delete for new collections.